packages/react-native/ReactCommon/react/renderer/core/RawValue.h

```diff
@@ -96,73 +111,157 @@ class RawValue {
    */
   template <typename T>
   explicit operator T() const {
-    return castValue(dynamic_, (T*)nullptr);
+    if (std::holds_alternative<folly::dynamic>(value_)) {
+      folly::dynamic dynamic = std::get<folly::dynamic>(value_);
+      return castValue(dynamic, (T*)nullptr);
+    } else {
+      const auto& [runtime, value] = std::get<JsiValuePair>(value_);
+      return castValue(runtime, value, (T*)nullptr);
+    }
   }
 
   inline explicit operator folly::dynamic() const noexcept {
-    return dynamic_;
+    return std::get<folly::dynamic>(value_);
   }
 
   /*
    * Checks if the stored value has specified type.
    */
   template <typename T>
-  bool hasType() const noexcept {
-    return checkValueType(dynamic_, (T*)nullptr);
+  bool hasType() const {
+    if (std::holds_alternative<folly::dynamic>(value_)) {
+      folly::dynamic dynamic = std::get<folly::dynamic>(value_);
+      return checkValueType(dynamic, (T*)nullptr);
+    } else {
+      const auto& [runtime, value] = std::get<JsiValuePair>(value_);
+      return checkValueType(runtime, value, (T*)nullptr);
+    }
   }
 
   /*
    * Checks if the stored value is *not* `null`.
    */
-  bool hasValue() const noexcept {
-    return !dynamic_.isNull();
+  bool hasValue() const {
+    if (std::holds_alternative<folly::dynamic>(value_)) {
+      folly::dynamic dynamic = std::get<folly::dynamic>(value_);
+      return !dynamic.isNull();
+    } else {
+      const auto& [runtime, value] = std::get<JsiValuePair>(value_);
+      return !value.isNull() && !value.isUndefined();
+    }
   }
 
  private:
-  folly::dynamic dynamic_;
+  using JsiValuePair = std::pair<jsi::Runtime*, jsi::Value>;
+  using ValueVariant = std::variant<folly::dynamic, JsiValuePair>;
```

Comment:
What is the shallow size of this structure compared to `dynamic` by itself?

Reply:
Variant is pretty compact - an info byte and whatever value is larger in memory afaik.. so I guess the size of the `jsi::Value` + the pointer + the info byte is roughly the same size if not smaller than `folly::dynamic`.

Reply:
A quick way to check this would be to put a `sizeof()` around it, then either hover if using something that has auto-completion, otherwise do a `static_assert` on it.

When I did some quick looks, `jsi::Value` after padding is the size of two pointers (one data, and one type). Then a pair of that and a pointer is three pointers. Then, adding variant bookkeeping, we get to four pointers (32 bytes) per prop is my guess?

These are allocated sparsely (only a RawValue for each present prop), compared to `Props` structures where we need to be more careful, since they scale to all available props. But something we should aim to still be mindful of since we keep `RawProps` around persistently for now.

Reply:
I'm wondering - is there any way we can lift out `jsi::Runtime` here? Then we skip the one pointer for Runtime, and the `std::pair` extra cost (if there is any, I assume it's 0 since it's just a templated `struct` afaik). Micro-optimization, but I guess it also simplifies the code a bit at the added risk of potentially mixing up different runtimes

Reply:
Forcing caller to have access to `jsi::runtime` might also be helpful from perspective of guaranteeing JS thread affinity requirement is well known.

It might just be a mess, but the other thing I wondered was having different types of RawValue, instead of making each RawValue store one of two types and bookkeep.

Reply:
The lifetime of RawValue is also limited compared to Props. We only use these while parsing, and then clear them out again.

Reply:
> But something we should aim to still be mindful of since we keep RawProps around persistently for now.

I don't think that's right. RawProps only exist for the duration of parsing / ShadowNode construction. 

We can also consider lifting this out in the future, right now it's a nice drop-in replacement.

Reply:
Yep, you’re right. Got this confused with rawProps Dynamic held on Android.